File: shop.c
// Project creating a shop in C
// Multi Paradigm Computing Module in Hiher Diploma in Data Analytics 2019
// Student Name: Yvonne Brady
// Student ID: G00376355


// Need to call these libraries in c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

// Create a container for the product information
struct Product {
	char* name;
	double price;
};

// Create a container for the stock information
struct ProductStock {
	struct Product product;
	int quantity;
};

// Create a container for the shop status
struct Shop {
	double cash;
	struct ProductStock stock[20];
	int index;
};

// Create a container for the customer information
struct Customer {
	char* name;
	double budget;
	struct ProductStock shoppingList[10];
	int index;
};

// Global Variables
// Easier to make them gloabal than passing them around everywhere
struct Shop shop;
struct Customer customer;

// Print info on product
void printProduct(struct Product p)
{
	printf("PRODUCT NAME: %s \nPRODUCT PRICE: EUR %.2f\n", p.name, p.price);
}


// Reads a csv file line by line to create stock list
struct Shop createAndStockShop()
{
    FILE * fp; // file
    char * line = NULL; // Each line of csv
    size_t len = 0;
    size_t read;

    fp = fopen("stock.csv", "r"); // Open the file as read only
    if (fp == NULL) // Error handling to exit if there is no file or cannot read
        exit(EXIT_FAILURE);

	// Reading in the initial shop float
	getline(&line, &len, fp);
	double cashInShop = atof(line); // convert it to a float
	shop.cash = cashInShop; // save it to the shop

	//This bit does the reading - line by line
    while ((read = getline(&line, &len, fp)) != -1) {
		//Reads the first item
		char *n = strtok(line, ","); // name
		// Null in this context means do not go on to the next item
		char *p = strtok(NULL, ","); // product
		char *q = strtok(NULL, ","); // quantity
		int quantity = atoi(q); // atoi converts the string into an integer
		double price = atof(p); // atof converts the string into a float
		char *name = malloc(sizeof(char) * 50); // manually allocating memory for this
		strcpy(name, n); // copy the properly foramtted / allocated string to the product name
		struct Product product = { name, price }; // write in the product
		struct ProductStock stockItem = { product, quantity }; // Add the quantities of each
		shop.stock[shop.index++] = stockItem; // index++ means that it starts as 0 and increases with each addition
		// This is a way to know the length of the array for looping around later. 
    }
	// returns the completed shop	
	return shop;
}

// Prints a stck sheet from the shop
void printShop()
{
	// Header info
	printf("\n***SHOP CONTENTS***\n\n");
	// How much cash is in the shop
	printf("FLOAT: EUR %.2f \n\n", shop.cash);
	// Loop through each stock item in the shop
	for (int i = 0; i < shop.index; i++)
	{
		// Call the printProduct function that prints the name and price
		printProduct(shop.stock[i].product);
		// Then print out the quantity in stock
		printf("QUANTITY IN STOCK: %d\n", shop.stock[i].quantity);
		printf("-------------\n"); // And a divider to make it pretty!
	}
}

// This is to check to see if we have enough stock to fulfill an order
// And updates quantities if we do.
bool checkProductStock(char *n, int order)
{
	// Loops through each stock item
	for (int i =0; i < shop.index; i++)
	{
		struct ProductStock product = shop.stock[i];
		// Gets the name from each item in stock
		char *name = product.product.name;
		// Compares it to the passed parameter product name
		if (strcmp(name, n) == 0) 
		{
			// Then checks if there is sufficient to fill the order
			if (shop.stock[i].quantity >= order){
				// Updates the shop stock to reflect the purchase
				shop.stock[i].quantity = shop.stock[i].quantity - order;
				// Lets the calling function know we have enough stock for the order
				return true;
			}
			else{
				// Otherwise lets the calling function know there is not enough stock for the order
				return false;
			}
		}
	}
	// Error - if the stock item was not found.
	return -1;
}

// This function finds the price of a product passed to it
double findProductPrice(char *n)
{
	// Loops through each of the shop stock items
	for (int i =0; i < shop.index; i++)
	{
		struct Product product = shop.stock[i].product;
		// Checks the product name against the name passed to the function
		char *name = product.name;
		// If there is a match
		if (strcmp(name, n) == 0) 
		{
			// return the price of that product to the calling function
			return product.price;
		}
	}
	// Otherwise return an erro value
	return -1;
}

// Reads a csv file line by line to create stock list
struct Customer readCustomer()
{
    FILE * fp; // file
    char * line = NULL; // Each line of csv
    size_t len = 0;
    size_t read;

    fp = fopen("customer.csv", "r"); // Open the file as read only
    if (fp == NULL) // Error handling to exit if there is no file or cannot read
        exit(EXIT_FAILURE);

	// Reading in the customer Name and Budget
	getline(&line, &len, fp);
	char *n = strtok(line, ","); // Customer Name
	char *c = strtok(NULL, ","); // Customer Budget
	double customerCash = atof(c); // Converting
	char *custName = malloc(sizeof(char) * 50); // Converting
	strcpy(custName, n); // Copying Name over
	// Creating the customer
	customer.name = custName;
	customer.budget = customerCash;
	struct ProductStock shopList;

	//This bit does the reading - line by line
    while ((read = getline(&line, &len, fp)) != -1) {
		//Reads the first item on a new item
		char *n = strtok(line, ","); // product name
		// Null in this context means don't go on to the next item
		char *q = strtok(NULL, ","); // quantity required
		int quantity = atoi(q); // atoi converts the string into an integer
		char *name = malloc(sizeof(char) * 50); // manually allocating memory for this
		strcpy(name, n); // again copying
		shopList.product.name = name; // assigning the variables - name
		shopList.product.price = findProductPrice(name); // assigning the variables - price
		shopList.quantity = quantity; // assigning the variables - quantity
		customer.shoppingList[customer.index++] = shopList; // index++ means that it starts as 0 and increases with each addition
		// This is a way to know the length of the array for looping around later. 
    }
	// And the customer is created	
	return customer;
}

// This is the livemode - it probably could be and possibly should be broken out into  smaller elements 0 but it works as is
void liveMode()
{
	// Declaring float variables for the customer budget and the total cost of the purchase
	double mybudget, totalCost;
	int quantity, select; // Integers for menu selection and product quantities
	// Need to ascertain the buget first
	printf("What is your budget in EURO (to two decimal places)?\n");
	scanf("%lf", &mybudget);
		// Keep bringin up the menu until told not to (99)
		do{
			// Banner
			printf("-----------------------------------\n   Welcome to the C Shop - LIVE MODE \n-----------------------------------\n");
			printf("\nHow can we help?\nYour current budget is %.2f.\nPlease select what you would like to buy from the list below:\n\n", mybudget);
			// Loop through every product and put them in a menu to see which is wanted to buy
			// Doing it this way reduces typing errors in product selection
			for(int i = 0; i < shop.index; i++)
				{
					// Menu selection per product
					printf("%d - %s @ %.2f each.\n", i+ 1, shop.stock[i].product.name, findProductPrice(shop.stock[i].product.name));
				}
				// And of course the menu out 
				printf("99 - Exit Live Mode\n");
				printf("\nPlease make your selection: ");
				// Get the selection
				scanf("%d", &select);
				// Do different things according to selection
				switch(select){
					// If you want out - bring out to earlier menu
					case 99:{
						printf("Thank you for choosing Live Mode. Please come again soon.\n");
					}	
					// Everything else - do this
					// This is the downside of looping through the prouct list - we don't knwo for sure what values are legitimate.
					default: {
						// Check to make sure it is within our limits
						if (select > shop.index+1) {
							printf ("Please enter a number from list above.\n");
							break;
						}
						// Assuming all ok - check the quantity required
						printf ("How many %ss would you like to purchase? ", shop.stock[select-1].product.name);
						scanf("%d", &quantity); 
						// Calculate the total cost
						totalCost = quantity * findProductPrice(shop.stock[select-1].product.name);
						// Check can they afford it
						if(mybudget < totalCost){
							// If not let them know
							printf("Insufficient funds for this transaction. No purchase made.\n");
						} else{
							// Check do we have enough stock. Here orders are either filled or not - not partial order of product
							if (checkProductStock(shop.stock[select-1].product.name, quantity)){
								// If all ok update the customer budget
								mybudget = mybudget - totalCost;
								// and the cash in the shop
								shop.cash = shop.cash + totalCost;
								// Let the customer know they were successful
								printf("%d %ss purchased at a total cost of EUR %.2f.\nNew budget is %.2f\n", quantity, shop.stock[select-1].product.name, totalCost, mybudget);
							}
							else{
								// Otherwise let them know there is not enough of the product
								printf("Insufficient stock to complete transaction. No purchase made.\n");
							}
						}
						// And the divider for clarity
						printf("---------------------------------------\n");
					}}}
				// Exit on 99 selection on menu
				while (select != 99);
}


// This prints out the shopping list and / or does the shopping via csv
// The boolean parameter passed in to the function represents whether this is just a read of the customer 
// or a read and update in the case of purcahsing via CSV
void printCustomer(bool upd)
{
	// Header info
	printf("\n***%s's SHOPPING LIST***\n", customer.name);
	// Setting the budget out
	printf("\nCUSTOMER NAME: %s \nCUSTOMER BUDGET: %.2f\n", customer.name, customer.budget);
	printf("-------------\n");
	// Loops through each of the items on the shopping list
	for(int i = 0; i < customer.index; i++)
		{
		// Calls the printProduct function which prints the product infornmtion
		printProduct(customer.shoppingList[i].product);
		// States how many are requested
		printf("QUANTITY REQUIRED: %d\n", customer.shoppingList[i].quantity);
		// Calculates the cost if the order was fulfilled
		double cost = customer.shoppingList[i].quantity * customer.shoppingList[i].product.price; 
		// Relays the information to the user
		printf("TOTAL ITEM COST: EUR %.2f\n", cost);

		// If this is a purchase as opposed to a read
		if (upd){
			// Check to see does the customer have enough funds
			if(customer.budget >= cost){
				// If they do - check to see if there is enough stock
				if (checkProductStock(customer.shoppingList[i].product.name, customer.shoppingList[i].quantity)){
					// If all is good update the customer budget to reflect the money thay have spent
					customer.budget = customer.budget - cost;
					// Give that money to the shop
					shop.cash = shop.cash + cost;
					// Take the quantity away from the shopplng list
					customer.shoppingList[i].quantity = 0;
					// Relay this info back to the customer / shop
					printf("QUANTITY PURCHASED: %d\n", customer.shoppingList[i].quantity);
					printf("ADJUSTED BUDGET: EUR %.2f\n", customer.budget);
					printf("ADJUSTED SHOP FLOAT: EUR %.2f\n", shop.cash);
					}
				else{
					// Let them know if there is not enough stock
					printf("Insufficient stock to fulfill order. No purchase made.\n");
				}}
			else{
				// Or if they do not have enough money
				printf("Insufficient funds to fulfill order. No purchase made.\n");
			}
		}
			// And the divider again
			printf("-------------\n");
	}
}

// This is the amin program
int main(void) 
{
	// Create the shop
	shop = createAndStockShop();
	// And the customer
	customer = readCustomer();
	// This variable is to record the user menu selection
	int menuSelect;
	// Keep showing the menu until the user wants out (99)
	do{
		// Header info
		printf("-------------------------------\n      Welcome to the C Shop\n-------------------------------\n");
		printf("\nHow can we help?\nPlease Select from the choices available below:\n\n");
		// Menu selection items
		printf("1 - Show current Shop contents\n"); // shows the current state of the shop
		printf("2 - Show current Shopping List\n"); // shows the current state of the customer shopping list
		printf("3 - Shop via CSV\n"); // Do a shop off the CSV
		printf("4 - Shop via Live Mode\n"); // Do a shop via live mode
		printf("5 - Reset Shop Quantities & Cash\n"); // Reset the shop to the original settings - this reads the csv again. The csv is not altered during the program.
		printf("6 - Reset Customer Quantities & Cash\n"); // Reset the customer similar to above
		printf("99 - Exit the Shop\n"); // Out
		printf("\nPlease make your selection:\n");
		scanf("%d", &menuSelect);
		switch (menuSelect){
			case 1:{
				// Show current shop contents - calls printShop above
				printShop();
				break;
			}
			case 2:{
				// Show shopping list- calls print Customer with the update bool set to false
				printCustomer(false);
				break;
			}
			case 3:{
				// Shop via CSV - same function as above but with update bool set to true
				printCustomer(true);
				break;
			}
			case 4:{
				// Live Mode
				printf("\nYou are now entering LIVE Mode!\n");
				liveMode();
				break;
			}
			case 5:{
				// Reset the shop stock and cash
				shop = createAndStockShop();
				printf("\n*** Shop now reset. ***\n");
				break;
			}
			case 6:{
				// Reset the customer shopping list and cash
				customer = readCustomer();
				printf("\n*** Customer now reset. ***!\n");
				break;
			}
			case 99:{
				// Personal exit message
				printf("\nGoodbye %s and thank you for your custom!\n", customer.name);
				break;
			}
			default:{
				// Incorrect entry message
				printf("\nIncorrect Selection - please try again\n");
				break;
			}
		}
	}
	// Continue until 99 is entered
	while (menuSelect != 99);
	// The end
    return 0;
}

File: JAVA/Shop/src/Shop/Shop.java
package Shop;

//Import all the packages needed 
import java.io.*; // system input and output through data streams, serialization and the file system
import java.nio.charset.StandardCharsets; // Defines charsets, decoders, and encoders, for translating between bytes and Unicode characters
import java.nio.file.Files; // for accessing files, file attributes, and file systems.
import java.nio.file.Paths; // for accessing files, file attributes, and file systems.
import java.util.ArrayList; // for ArrayLists (ShoppingList is ArrayList of ProductStock
import java.util.Collections; // static methods that operate on or return collections.
import java.util.List; // looks after lists
import java.util.Scanner; // for user inputs

public class Shop {
	
	// Variables used to define the shop
	private double cash;
	private ArrayList<ProductStock> stock;

	public Shop(String fileName) {
		// The constructor
		stock = new ArrayList<>();
		// List to read the csv lines into
		List<String> lines = Collections.emptyList();
		
		// try - to provide some error handling if there is a problem with the file
		try {
			// Read each line into the lines list
			lines = Files.readAllLines(Paths.get(fileName), StandardCharsets.UTF_8);
			// Cash is in the first line
			cash = Double.parseDouble(lines.get(0));
			// Removing first line as it is the only non-stock item
			lines.remove(0);
			// For every line in your file
			for (String line : lines) {
				// Define the comma as the delimiter between values
				String[] arr = line.split(",");
				// Product Name, Price and Quantity values separated 
				String name = arr[0];
				double price = Double.parseDouble(arr[1]);
				int quantity = Integer.parseInt(arr[2].trim()); // trim to get rid of whitespace
				// Map to the shop variables
				Product p = new Product(name, price);
				ProductStock s = new ProductStock(p, quantity);
				// Add to the shops stock
				stock.add(s);
			}
		}

		catch (IOException e) {
			// if file is not found or unavailable
			System.out.println("Error accessing " + fileName + " file. Please ensure this file exists and is accessible.");
			System.out.println("Quitting shop. Goodbye!");
			System.exit(0);
		}
	}

	// Getters
	public double getCash() {
		return cash;
	}

	public ArrayList<ProductStock> getStock() {
		return stock;
	}

	// Print out the shop contents
	@Override
	public String toString() {
		return ("\n*** SHOP CONTENTS ***\n\nFLOAT: EUR " + String.format("%.2f", cash) + "\n\n" + stock + "\n").replace("[","").replace("]","").replace(", ", "");
		
	}

	// Finds the product from the stock list - used to retrieve price and quantity information 
	// Takes in the product name
	private ProductStock findProduct(String name) {
		// Loops through the stock list
		for (ProductStock ps : stock) {
			// Gets the product info
			Product p = ps.getProduct();
			// Compares the product names
			if (p.getName().equals(name)) {
				// If they match - return the product stock info
				return ps;
			}
		}
		// If the product is not found
		Product p = new Product("error", -1.0); 
		ProductStock ps = new ProductStock(p,0);
		return ps;
	}
	
	// Adds the prices to the shopping list
	public void applyPrices(Customer c) {
		// Loops through each item on the shopping list
		for (ProductStock productStock : c.getShoppingList()) {
			// Get the product from the product stock
			Product p = productStock.getProduct();
			// Get the shops Product Stock info on the product and then its price  
			double price = findProduct(p.getName()).getProduct().getPrice();
			// Set the price in the shopping list
			p.setPrice(price);
		}
	}
	
	// This is the method use to perform the transaction
	public void processOrder(Customer c) {
		// If it is in Shopping List mode - print some header info
		if (c.getName() != "Live Mode") {
			String head = "\n*** "+ c.getName().toUpperCase() + "'S SHOPPING LIST ***\n";
			head += "\nCUSTOMER NAME: " + c.getName();
			head +="\nCUSTOMER BUDGET:" + String.format("%.2f",c.getBudget());
			head += "\n-------------";
			System.out.println(head);
		}
		// ret used to build a string with the output
		String ret;
		// Loop through was shopping list item
		// In the case of live mode there will only be one item on the list
		for (ProductStock cListItem : c.getShoppingList()) {
			// Determine the quantity required
			int qr = cListItem.getQuantity();
			// The product required - used throughout so better to make it a parameter
			Product prodReq = cListItem.getProduct();
			// Get the price of the product
			double pr = prodReq.getPrice();
			// The product name
			String pname = prodReq.getName();
			// Find the associated product stock item inthe shop
			ProductStock shopProd = findProduct(pname);
			// calculate the cost of the purchase
			double cost = pr*qr;
			// Create a new parameter for the quantity actually bought. 
			// This is set to the quantity required initially 
			// but may be revised if the full order cannot be met.
			int qb = qr; // quantity bought - assumed initially to be the amount requested
			// Start building the ret string
			ret = "PRODUCT NAME: " + prodReq.getName();
			ret +="\nPRODUCT PRICE: EUR "+ String.format("%.2f", pr);
			ret += "\nQUANTITY REQUIRED: " + qr;
			
			if (pr > -1) {
				// Checks to see if the user has enough money
				if (c.getBudget() < cost){
					// If they don't have enough money, set the quantity bought to what they can afford
					// Dividing two ints will result in the quotient returned so no other cals required.
					qb = (int)(c.getBudget() / pr);
				}
				
				// Checks to see if the shop has enough stock
				if(qb >  shopProd.getQuantity()) {
					// If not set the quantity bought to the quantity remaining in shop.
					qb = shopProd.getQuantity();
				}
				
				// Adjust cost accordingly
				cost = (pr*qb);
				
				// Update the customer cash and quantity details to reflect the purchases made 
				cListItem.setQuantity(qr-qb);
				c.setBudget(c.getBudget()-cost);
				
				// Update the shop float and stock details
				shopProd.setQuantity(shopProd.getQuantity()-qb);
				cash += cost;
				
				// Continue to build the report out string with the details
				ret += "\nQUANTITY PURCHASED: " + qb;
				ret += "\nTOTAL ITEM COST: EUR "+ String.format("%.2f", cost);
				ret += "\nADJUSTED BUDGET: EUR " + String.format("%.2f",c.getBudget());
				ret += "\nADJUSTED SHOP FLOAT: EUR " + String.format("%.2f",cash);
				// If the order was not fulfilled completely
				if(qb != qr) {
					// Inform the customer
					ret +="\nUnable to fulfill complete order on this item due to insufficient funds / stock.";
				}
			}
			else {
				// If product is not in stock list
				ret+= "\nSorry, shop does not stock " + prodReq.getName();
			}
			ret += "\n-------------";
			System.out.println(ret);
				
			}
	}
	
	// Method to perform the live mode
	public void liveMode() {
		// New scanner - this is not closed as to do so would also close the main menu scanner. 
		// Both are closed on exiting from the main menu.
		Scanner liveScan = new Scanner(System.in);
		// Header info
		System.out.println();		
		System.out.println("You are now entering Live Mode!");
		System.out.println("What is your budget in EURO (to two decimal places)?");
		
		// Record the customer budget
		double liveBudget = liveScan.nextDouble();
		// To track the cutomers menu choice 
		int liveChoice;
		
		// Keep repeating the menu until the customer exits live mode
	    do {
			// Header info
			System.out.println("\n-------------------------------\n Welcome to the JAVA Shop - LIVE MODE \n-------------------------------");
			System.out.println("How can we help?\nYour current budget is EUR "+ String.format("%.2f",liveBudget));
			System.out.println("Please select what you would like to buy from the list below:\n");
			
			// Used to keep track of choices
			int i = 1;
			// Menu selection items - dynamically built from product stock
			// Loop through each stock item
			for (ProductStock ps : stock) {
				Product p = ps.getProduct();
				// Add that item as a menu choice together with the price of the product
				System.out.println(i + " - " + p.getName() + " @ " + String.format("%.2f",p.getPrice())+ " each.");
				i+=1;
			}
			// Add the option to exit and instructions to make the choice
			System.out.println("99 - Exit Live Mode"); // Out
			System.out.println("\nPlease make your selection:");
	    	
			// Wait until an integer has been entered
			liveChoice = liveScan.nextInt();
			// If it is a valid choice
			if (liveChoice <i) {
				// Record the product name as a variable
				String prodName = stock.get(liveChoice-1).getProduct().getName();
				// See how many the customer would like to purchase?
				System.out.println("How many "+ prodName + "s would you like to purchase?");
				// Record the quantity required
				int liveQuantity = liveScan.nextInt();
				// Find the price of the product as before
				double livePrice = findProduct(prodName).getProduct().getPrice();
				
				// Put that info into product and product stock formats
				Product liveProd = new Product(prodName,livePrice);
				ProductStock liveItem = new ProductStock(liveProd,liveQuantity);
				// Create the new customer with the live details
				Customer liveCust = new Customer("Live Mode", liveBudget, liveItem);

				// Perform the transaction as above
				processOrder(liveCust);
				
				// Update the budget after purchase
				liveBudget = liveCust.getBudget();
			}
				// On exit live mode
			else if (liveChoice == 99) {	    
				System.out.println("Thank you for chosing LIVE mode. Please come again soon.");

				}
				// Otherwise - an error message
			else {
				System.out.println("Please enter a valid selection");
			}
	       // Exit on 99
	    } while (liveChoice != 99);
	}
}